flink-autoscaler/src/main/java/org/apache/flink/autoscaler/JobVertexScaler.java

```diff
@@ -378,28 +405,70 @@ protected static int scale(
 
         // Cap parallelism at either maxParallelism(number of key groups or source partitions) or
         // parallelism upper limit
-        final int upperBound = Math.min(maxParallelism, parallelismUpperLimit);
+        int upperBound = Math.min(maxParallelism, parallelismUpperLimit);
 
         // Apply min/max parallelism
         newParallelism = Math.min(Math.max(parallelismLowerLimit, newParallelism), upperBound);
 
         var adjustByMaxParallelism =
                 inputShipStrategies.isEmpty() || inputShipStrategies.contains(HASH);
         if (!adjustByMaxParallelism) {
-            return newParallelism;
+            return Tuple2.of(newParallelism, Optional.empty());
         }
 
-        // When the shuffle type of vertex inputs contains keyBy or vertex is a source, we try to
-        // adjust the parallelism such that it divides the maxParallelism without a remainder
-        // => data is evenly spread across subtasks
-        for (int p = newParallelism; p <= maxParallelism / 2 && p <= upperBound; p++) {
-            if (maxParallelism % p == 0) {
-                return p;
+        if (numPartitions <= 0) {
+            // When the shuffle type of vertex inputs contains keyBy or vertex is a source,
+            // we try to adjust the parallelism such that it divides the maxParallelism without a
+            // remainder => data is evenly spread across subtasks
+            for (int p = newParallelism; p <= maxParallelism / 2 && p <= upperBound; p++) {
+                if (maxParallelism % p == 0) {
+                    return Tuple2.of(p, Optional.empty());
+                }
+            }
+            // If parallelism adjustment fails, use originally computed parallelism
+            return Tuple2.of(newParallelism, Optional.empty());
+        } else {
+
+            // When we know the numPartitions at a vertex,
+            // adjust the parallelism such that it divides the numPartitions without a remainder
+            // => Data is evenly distributed among subtasks
+            for (int p = newParallelism; p <= upperBound && p <= numPartitions; p++) {
+                if (numPartitions % p == 0) {
+                    return Tuple2.of(p, Optional.empty());
+                }
             }
```

Comment:
This looks like code duplication to me. Only the for loop termination condition changed. We should be able to pass an argument to the for loop which we set based on the number of partitions.

We may even completely simplify like this:

```suggestion
        if (numPartitions <= 0) {
            // No partition information is available, assume numPartitions equals the number of key groups
            numPartitions = maxParallelism;
        }
       for (int p = newParallelism; p <= maxParallelism / 2 && p <= upperBound; p++) {
            if (maxParallelism % p == 0) {
                return Tuple2.of(p, Optional.empty());
            }
        }
        // If parallelism adjustment fails, use originally computed parallelism
        return Tuple2.of(newParallelism, Optional.empty());
```

Reply:
> Right, I missed that. I was trying to generalize the two code blocks. How about the following?

Thansk, Fine with me , However, I do not recommend overriding the values ​​​​of MaxParallelism and UpperBound, so I added two new variables instead:

1. adjustableMaxParallelism( Indicates the MaxParallelism  in the adjustment process )
2. adjustableUpperBound (Indicates the UpperBound in the adjustment process)

Reply:
Can we call them like this?

1. adjustableMaxParallelism => numKeyGroupsOrPartitions
2. adjustableUpperBound =>  upperBoundForAlignment  

Adjustable just doesn't tell someone who is unfamiliar with the code very much.